gateway/mw_api_rate_limit.go

```diff
@@ -24,8 +25,59 @@ func (k *RateLimitForAPI) Name() string {
 	return "RateLimitForAPI"
 }
 
+func (k *RateLimitForAPI) shouldEnable() bool {
+	if k.Spec.DisableRateLimit {
+		return false
+	}
+
+	// per endpoint rate limits
+	for _, version := range k.Spec.VersionData.Versions {
+		for _, v := range version.ExtendedPaths.RateLimit {
+			if !v.Disabled {
+				return true
+			}
+		}
+	}
+
+	// global api rate limit
+	if k.Spec.GlobalRateLimit.Rate == 0 || k.Spec.GlobalRateLimit.Disabled {
+		return false
+	}
+
+	return true
+}
+
+func (k *RateLimitForAPI) getSession(r *http.Request) *user.SessionState {
+	versionInfo, _ := k.Spec.Version(r)
+	versionPaths := k.Spec.RxPaths[versionInfo.Name]
+
+	found, meta := k.Spec.CheckSpecMatchesStatus(r, versionPaths, RateLimit)
+	if found {
+		limits, ok := meta.(*apidef.RateLimitMeta)
+		if !ok {
+			log.Fatalf("Expected RateLimitMeta, got %T", meta)
```

Comment:
`log.Fatalf` would exit the application. This should be logged with `ERROR` severity and return error

Reply:
This branch can only be triggered in development, it would fail only if a developer renames the RateLimitMeta type or moves it outside of apidef. I feel there is a way to remove this assertion altogether, however we'd need more api adjustments of CheckSpecMatchesStatus:

```
		case RateLimit:
			if method == rxPaths[i].RateLimit.Method {
				return true, &rxPaths[i].RateLimit
			}
```

I'd suggest to focus on CheckSpecMatchesStatus to decompose it's responsibilities:

- find which spec by method/path of request: (*URLSpec, bool) - no "subset" of URLSpec
- reference urlspec.RateLimit without needing to type cast

Essentially the function we need should just return either `i` or `&rxPaths[i]` (*URLSpec), and we avoid loosing type safety which is the real issue.

Reply:
Rewrote CheckSpecMatchesStatus for type safety, added FindSpecMatchesStatus. This gives us a cleanup to do removing the old function. FindSpecMatchesStatus is used to get the fully matched *URLSpec, and then a type safe reference to the field is used from the middleware.